### PR TITLE
Add DigitalOcean Droplet Routes (Take 2)

### PR DESCRIPTION
**Pull request by andrewgwoodruff**
This MR adds the various DigitalOcean droplet routes, utilizing the open-realms/do-minecraft library. This addresses https://github.com/open-realms/command-block/issues/2.

### CODE REVIEW

File: src/droplet/droplet.controller.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+export class DropletController {
+  constructor(private readonly dropletService: DropletService) {}
+
+  @Get()
+  getAllDroplets(): string {
+    return this.dropletService.getAllDroplets();
+  }
+
+  @Post()
+  createDroplet(): string {
```
Comment (johnbwoodruff) [inline src/droplet/droplet.controller.ts:14]:
```suggestion
  async createDroplet(): string {
```

Comment (johnbwoodruff) [inline src/droplet/droplet.controller.ts]:
```suggestion
  async createDroplet(): string {
```

File: src/droplet/droplet.controller.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    return this.dropletService.createDroplet();
```
Comment (johnbwoodruff) [inline src/droplet/droplet.controller.ts]:
```suggestion
    return await this.dropletService.createDroplet();
```

File: src/droplet/droplet.controller.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  @Delete(':id')
+  deleteDroplet(@Param() params): string {
```
Comment (johnbwoodruff) [inline src/droplet/droplet.controller.ts]:
```suggestion
  async deleteDroplet(@Param() params): string {
```

File: src/droplet/droplet.controller.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    return this.dropletService.deleteDroplet(params.id);
```
Comment (johnbwoodruff) [inline src/droplet/droplet.controller.ts]:
```suggestion
    return await this.dropletService.deleteDroplet(params.id);
```

File: src/droplet/droplet.controller.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  @Get(':id')
+  getDropletDetails(@Param() params): string {
```
Comment (johnbwoodruff) [inline src/droplet/droplet.controller.ts]:
```suggestion
  async getDropletDetails(@Param() params): string {
```

File: src/droplet/droplet.controller.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    return this.dropletService.getDropletDetails(params.id);
```
Comment (johnbwoodruff) [inline src/droplet/droplet.controller.ts]:
```suggestion
    return await this.dropletService.getDropletDetails(params.id);
```

File: src/droplet/droplet.controller.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  @Put(':id')
+  updateDropletDetails(@Param() params): string {
```
Comment (johnbwoodruff) [inline src/droplet/droplet.controller.ts]:
```suggestion
  async updateDropletDetails(@Param() params): string {
```

File: src/droplet/droplet.controller.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    return this.dropletService.updateDropletDetails(params.id);
```
Comment (johnbwoodruff) [inline src/droplet/droplet.controller.ts]:
```suggestion
    return await this.dropletService.updateDropletDetails(params.id);
```

File: src/droplet/droplet.controller.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  @Put(':id/start')
+  startDroplet(@Param() params): string {
```
Comment (johnbwoodruff) [inline src/droplet/droplet.controller.ts]:
```suggestion
  async startDroplet(@Param() params): string {
```

File: src/droplet/droplet.controller.ts [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    return this.dropletService.startDroplet(params.id);
```
Comment (johnbwoodruff) [inline src/droplet/droplet.controller.ts]:
```suggestion
    return await this.dropletService.startDroplet(params.id);
```

File: src/droplet/droplet.controller.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  @Put(':id/stop')
+  stopDroplet(@Param() params): string {
```
Comment (johnbwoodruff) [inline src/droplet/droplet.controller.ts]:
```suggestion
  async stopDroplet(@Param() params): string {
```

File: src/droplet/droplet.controller.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    return this.dropletService.stopDroplet(params.id);
```
Comment (johnbwoodruff) [inline src/droplet/droplet.controller.ts]:
```suggestion
    return await this.dropletService.stopDroplet(params.id);
```

File: src/droplet/droplet.service.ts [hunk 13]
```diff
@@ -0,0 +1,32 @@
+import { Injectable } from '@nestjs/common';
+
+@Injectable()
+export class DropletService {
+  getAllDroplets(): string {
```
Comment (johnbwoodruff) [inline src/droplet/droplet.service.ts]:
```suggestion
  async getAllDroplets(): string {
```

File: src/droplet/droplet.service.ts [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    return 'Here are all the servers!';
+  }
+
+  createDroplet(): string {
```
Comment (johnbwoodruff) [inline src/droplet/droplet.service.ts]:
```suggestion
  async createDroplet(): string {
```

File: src/droplet/droplet.service.ts [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+    return 'A new server has been created!';
+  }
+
+  deleteDroplet(id: string): string {
```
Comment (johnbwoodruff) [inline src/droplet/droplet.service.ts]:
```suggestion
  async deleteDroplet(id: string): string {
```

File: src/droplet/droplet.service.ts [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+    return 'Server ${id} has been deleted!';
+  }
+
+  getDropletDetails(id: string): string {
```
Comment (johnbwoodruff) [inline src/droplet/droplet.service.ts]:
```suggestion
  async getDropletDetails(id: string): string {
```

File: src/droplet/droplet.service.ts [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+    return 'Here are the details for this server: ${id}';
+  }
+
+  updateDropletDetails(id: string): string {
```
Comment (johnbwoodruff) [inline src/droplet/droplet.service.ts]:
```suggestion
  async updateDropletDetails(id: string): string {
```

File: src/droplet/droplet.service.ts [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+    return 'Server ${id} has been updated!';
+  }
+
+  startDroplet(id: string): string {
```
Comment (johnbwoodruff) [inline src/droplet/droplet.service.ts]:
```suggestion
  async startDroplet(id: string): string {
```

File: src/droplet/droplet.service.ts [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+    return 'Server ${id} has been started!';
+  }
+
+  stopDroplet(id: string): string {
```
Comment (johnbwoodruff) [inline src/droplet/droplet.service.ts]:
```suggestion
  async stopDroplet(id: string): string {
```

File: src/droplet/droplet.controller.spec.ts [hunk 20]
```diff
@@ -0,0 +1,18 @@
+import { Test, TestingModule } from '@nestjs/testing';
+import { DropletController } from './droplet.controller';
+
+describe('DropletController', () => {
```
Comment (johnbwoodruff) [inline src/droplet/droplet.controller.spec.ts]: I'd love for us to start out doing tests from the beginning of this project.

**Review by johnbwoodruff — CHANGES_REQUESTED** [review]

**Comment by andrewgwoodruff** [comment]
Just to make sure I understand, the `:id/start` and `:id/stop` PUT endpoints are just abstractions as far as the frontend is concerned, yes? Like an end user will have the option to `stop` their world, and when they click that button, what actually happens on the backend is that we spin down the server, take a snapshot, and kill the DigitalOcean droplet. Is that correct? Likewise, when the end-user hits the `start world` button, it does a PUT request to the `:id/start` endpoint, which fetches the existing snapshot and uses that to create a new droplet with that snapshot. Does that reflect your expectation of how those two endpoints will work, John?

If so, that would mean that the DELETE `:id` endpoint goes and actually kills/deletes/destroys the whole droplet and doesn't save anything. Let me know if that's what you're going for.

**Comment by johnbwoodruff** [comment]
Closed in favor of #10

**Pull request closed without merging**